users/views.py

```diff
@@ -110,40 +115,44 @@ def login_process(request):
 
 def signup_view(request):
 	form = SignupForm()
-	context = {'form': form, 'result': request.GET.get('result')}
+	context = generate_signup_form(form)
+	context['result'] = request.GET.get('result')
 	return render(request, 'users/signup.html', context)
 
+def generate_signup_form(form):
+	(question, answer, chelp) = makeCaptcha()
+	return {'form': form, 'captcha': question, 'token': hash2(answer), 'captcha_help': chelp}
+
+def hash2(inp):
+	return sha256(str(inp).encode()).hexdigest()
+
 def signup_process(request):
 	form = SignupForm(request.POST)
 	if(form.is_valid()):
 
 		data_valid = True
+		captcha_input = re.sub(r'\s','',request.POST['captcha'])
+		captcha_target = re.sub(r'\s','',request.POST['captcha-token'])
 
-		# Validate captcha
-		import requests
-		import os
-		data = {
-			'secret': os.environ['GCAPTCHA_SECRET'],
-			'response': request.POST['g-recaptcha-response'],
-		}
-		r = requests.post('https://www.google.com/recaptcha/api/siteverify', data=data)
-
-		if not r.json()['success']:
-			form.add_error('g-recaptcha-response', ValidationError('Captcha failed.'))
-			data_valid = False
+		#print(hash2(captcha_input), captcha_input, captcha_target)
 
+		# CHECK CAPTCHA!
+		if hash2(captcha_input) != captcha_target:
+			# Should be attached to captcha, but this is the closest thing.
+			form.add_error('username', ValidationError('You didn\'t answer the captcha correctly!'))
+			data_valid = False
 		# CHECK CASE!
-		if User.objects.filter(username__iexact=form.cleaned_data['username']).exists():
+		elif User.objects.filter(username__iexact=form.cleaned_data['username']).exists():
 			form.add_error('username', ValidationError('A user with that name already exists.'))
 			data_valid = False
-		
+
 		if not data_valid:
-			return render(request, 'users/signup.html', {'form': form})
+			return render(request, 'users/signup.html', generate_signup_form(form))
```

Comment:
is it intentional that the context has changed from `{'form': form}` to `form`
and does the template reflect this change

Reply:
never mind, i just looked back at `generate_signup_form`.
seems odd to have a function that takes a form, and both adds a captcha and puts it in an object?

Reply:
Renamed as per earlier. :pencil2: